Hazelcast fails to start when k8s does not have CoreDNS service enabled

I am running Hazelcast 4.1-BETA-1 in [microk8s.io](https://microk8s.io/) cluster in a default configuration. Hazelcast correctly detects it's running in k8s, but it fails to start:
```
########################################
# JAVA_OPTS=-Djava.net.preferIPv4Stack=true -Djava.util.logging.config.file=/opt/hazelcast/logging.properties -XX:MaxRAMPercentage=80.0 -XX:+UseParallelGC --add-modules java.se --add-exports java.base/jdk.internal.ref=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.nio=ALL-UNNAMED --add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.management/sun.management=ALL-UNNAMED --add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED
# CLASSPATH=/opt/hazelcast/*:/opt/hazelcast/lib/*
# starting now....
+ exec java -server -Djava.net.preferIPv4Stack=true -Djava.util.logging.config.file=/opt/hazelcast/logging.properties -XX:MaxRAMPercentage=80.0 -XX:+UseParallelGC --add-modules java.se --add-exports java.base/jdk.internal.ref=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.nio=ALL-UNNAMED --add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.management/sun.management=ALL-UNNAMED --add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED com.hazelcast.core.server.HazelcastMemberStarter
########################################
Oct 31, 2020 7:44:16 PM com.hazelcast.internal.config.AbstractConfigLocator
INFO: Loading 'hazelcast-default.xml' from the classpath.
Oct 31, 2020 7:44:20 PM com.hazelcast.system
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Hazelcast 4.1-BETA-1 (20200909 - f3c5b19) starting at [10.1.243.195]:5701
Oct 31, 2020 7:44:23 PM com.hazelcast.spi.discovery.integration.DiscoveryService
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Auto-detection selected discovery strategy: class com.hazelcast.kubernetes.HazelcastKubernetesDiscoveryStrategyFactory
Oct 31, 2020 7:44:23 PM com.hazelcast.spi.discovery.integration.DiscoveryService
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Kubernetes Discovery properties: { service-dns: null, service-dns-timeout: 5, service-name: null, service-port: 0, service-label: null, service-label-value: true, namespace: default, pod-label: null, pod-label-value: null, resolve-not-ready-addresses: true, use-node-name-as-external-address: false, kubernetes-api-retries: 3, kubernetes-master: https://kubernetes.default.svc}
Oct 31, 2020 7:44:23 PM com.hazelcast.spi.discovery.integration.DiscoveryService
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Kubernetes Discovery activated with mode: KUBERNETES_API
Oct 31, 2020 7:44:24 PM com.hazelcast.instance.impl.Node
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Using Discovery SPI
Oct 31, 2020 7:44:24 PM com.hazelcast.cp.CPSubsystem
WARNING: [10.1.243.195]:5701 [dev] [4.1-BETA-1] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
Oct 31, 2020 7:44:25 PM com.hazelcast.internal.diagnostics.Diagnostics
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
Oct 31, 2020 7:44:25 PM com.hazelcast.core.LifecycleService
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] [10.1.243.195]:5701 is STARTING
Oct 31, 2020 7:44:32 PM com.hazelcast.kubernetes.RetryUtils
WARNING: Couldn't discover Hazelcast members using Kubernetes API, [1] retrying in 1 seconds...
Oct 31, 2020 7:44:34 PM com.hazelcast.kubernetes.RetryUtils
WARNING: Couldn't discover Hazelcast members using Kubernetes API, [2] retrying in 2 seconds...
Oct 31, 2020 7:44:36 PM com.hazelcast.kubernetes.RetryUtils
WARNING: Couldn't discover Hazelcast members using Kubernetes API, [3] retrying in 3 seconds...
Oct 31, 2020 7:44:39 PM com.hazelcast.spi.discovery.integration.DiscoveryService
WARNING: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Cannot fetch the current zone, ZONE_AWARE feature is disabled
Oct 31, 2020 7:44:39 PM com.hazelcast.kubernetes.RetryUtils
WARNING: Couldn't discover Hazelcast members using Kubernetes API, [1] retrying in 1 seconds...
Oct 31, 2020 7:44:41 PM com.hazelcast.kubernetes.RetryUtils
WARNING: Couldn't discover Hazelcast members using Kubernetes API, [2] retrying in 2 seconds...
Oct 31, 2020 7:44:48 PM com.hazelcast.kubernetes.RetryUtils
WARNING: Couldn't discover Hazelcast members using Kubernetes API, [3] retrying in 3 seconds...
Oct 31, 2020 7:44:51 PM com.hazelcast.internal.cluster.impl.DiscoveryJoiner
SEVERE: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Failure in executing REST call
com.hazelcast.kubernetes.RestClientException: Failure in executing REST call
	at com.hazelcast.kubernetes.RestClient.call(RestClient.java:115)
	at com.hazelcast.kubernetes.RestClient.get(RestClient.java:80)
	at com.hazelcast.kubernetes.KubernetesClient$1.call(KubernetesClient.java:496)
	at com.hazelcast.kubernetes.KubernetesClient$1.call(KubernetesClient.java:490)
	at com.hazelcast.kubernetes.RetryUtils.retry(RetryUtils.java:52)
	at com.hazelcast.kubernetes.KubernetesClient.callGet(KubernetesClient.java:490)
	at com.hazelcast.kubernetes.KubernetesClient.endpoints(KubernetesClient.java:81)
	at com.hazelcast.kubernetes.KubernetesApiEndpointResolver.resolve(KubernetesApiEndpointResolver.java:70)
	at com.hazelcast.kubernetes.HazelcastKubernetesDiscoveryStrategy.discoverNodes(HazelcastKubernetesDiscoveryStrategy.java:108)
	at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.discoverNodes(DefaultDiscoveryService.java:72)
	at com.hazelcast.internal.cluster.impl.DiscoveryJoiner.getPossibleAddresses(DiscoveryJoiner.java:69)
	at com.hazelcast.internal.cluster.impl.DiscoveryJoiner.getPossibleAddressesForInitialJoin(DiscoveryJoiner.java:58)
	at com.hazelcast.internal.cluster.impl.TcpIpJoiner.joinViaPossibleMembers(TcpIpJoiner.java:136)
	at com.hazelcast.internal.cluster.impl.TcpIpJoiner.doJoin(TcpIpJoiner.java:96)
	at com.hazelcast.internal.cluster.impl.AbstractJoiner.join(AbstractJoiner.java:137)
	at com.hazelcast.instance.impl.Node.join(Node.java:808)
	at com.hazelcast.instance.impl.Node.start(Node.java:471)
	at com.hazelcast.instance.impl.HazelcastInstanceImpl.<init>(HazelcastInstanceImpl.java:123)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.constructHazelcastInstance(HazelcastInstanceFactory.java:211)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:190)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:128)
	at com.hazelcast.core.Hazelcast.newHazelcastInstance(Hazelcast.java:91)
	at com.hazelcast.core.server.HazelcastMemberStarter.main(HazelcastMemberStarter.java:47)
Caused by: java.net.UnknownHostException: kubernetes.default.svc
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:220)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:403)
	at java.base/java.net.Socket.connect(Socket.java:609)
	at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:285)
	at java.base/sun.security.ssl.BaseSSLSocketImpl.connect(BaseSSLSocketImpl.java:173)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:182)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
	at java.base/sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:265)
	at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:372)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:191)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:177)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1587)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1515)
	at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:527)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:334)
	at com.hazelcast.kubernetes.RestClient.checkHttpOk(RestClient.java:132)
	at com.hazelcast.kubernetes.RestClient.call(RestClient.java:112)
	... 22 more

Oct 31, 2020 7:44:51 PM com.hazelcast.instance.impl.Node
SEVERE: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Could not join cluster. Shutting down now!
Oct 31, 2020 7:44:51 PM com.hazelcast.core.LifecycleService
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] [10.1.243.195]:5701 is SHUTTING_DOWN
Oct 31, 2020 7:44:51 PM com.hazelcast.instance.impl.Node
WARNING: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Terminating forcefully...
Oct 31, 2020 7:44:51 PM com.hazelcast.instance.impl.Node
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Shutting down connection manager...
Oct 31, 2020 7:44:51 PM com.hazelcast.instance.impl.Node
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Shutting down node engine...
Oct 31, 2020 7:44:51 PM com.hazelcast.instance.impl.NodeExtension
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Destroying node NodeExtension.
Oct 31, 2020 7:44:51 PM com.hazelcast.instance.impl.Node
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Hazelcast Shutdown is completed in 61 ms.
Oct 31, 2020 7:44:51 PM com.hazelcast.core.LifecycleService
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] [10.1.243.195]:5701 is SHUTDOWN
Oct 31, 2020 7:44:52 PM com.hazelcast.internal.util.PhoneHome
WARNING: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Could not schedule phone home task! Most probably Hazelcast failed to start.
Oct 31, 2020 7:44:52 PM com.hazelcast.instance.impl.Node
INFO: [10.1.243.195]:5701 [dev] [4.1-BETA-1] Node is already shutting down... Waiting for shutdown process to complete...
Exception in thread "main" java.lang.IllegalStateException: Node failed to start!
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
	at com.hazelcast.internal.util.ExceptionUtil.tryWrapInSameClass(ExceptionUtil.java:233)
	at com.hazelcast.internal.util.ExceptionUtil.wrapException(ExceptionUtil.java:138)
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:114)
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:73)
	at com.hazelcast.internal.util.ExceptionUtil.rethrow(ExceptionUtil.java:148)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:194)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:128)
	at com.hazelcast.core.Hazelcast.newHazelcastInstance(Hazelcast.java:91)
	at com.hazelcast.core.server.HazelcastMemberStarter.main(HazelcastMemberStarter.java:47)
Caused by: java.lang.IllegalStateException: Node failed to start!
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
	at com.hazelcast.internal.util.ExceptionUtil.tryWrapInSameClass(ExceptionUtil.java:233)
	at com.hazelcast.internal.util.ExceptionUtil.wrapException(ExceptionUtil.java:138)
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:114)
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:73)
	at com.hazelcast.internal.util.ExceptionUtil.rethrow(ExceptionUtil.java:148)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.constructHazelcastInstance(HazelcastInstanceFactory.java:237)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:190)
	... 3 more
Caused by: java.lang.IllegalStateException: Node failed to start!
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
	at com.hazelcast.internal.util.ExceptionUtil.tryWrapInSameClass(ExceptionUtil.java:233)
	at com.hazelcast.internal.util.ExceptionUtil.wrapException(ExceptionUtil.java:138)
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:114)
	at com.hazelcast.internal.util.ExceptionUtil.peel(ExceptionUtil.java:73)
	at com.hazelcast.internal.util.ExceptionUtil.rethrow(ExceptionUtil.java:148)
	at com.hazelcast.instance.impl.HazelcastInstanceImpl.<init>(HazelcastInstanceImpl.java:143)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.constructHazelcastInstance(HazelcastInstanceFactory.java:211)
	... 4 more
Caused by: java.lang.IllegalStateException: Node failed to start!
	at com.hazelcast.instance.impl.HazelcastInstanceImpl.<init>(HazelcastInstanceImpl.java:125)
	... 5 more
```

Apparently it's because [microk8s does not enable CoreDNS](https://microk8s.io/docs/addons) by default. I am not sure whether we can do anything about it.

--- a/src/main/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategyFactory.java
+++ b/src/main/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategyFactory.java
@@ -18,12 +18,15 @@ package com.hazelcast.kubernetes;
 
 import com.hazelcast.config.properties.PropertyDefinition;
 import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
 import com.hazelcast.spi.discovery.DiscoveryNode;
 import com.hazelcast.spi.discovery.DiscoveryStrategy;
 import com.hazelcast.spi.discovery.DiscoveryStrategyFactory;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 
 import java.io.File;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
@@ -34,6 +37,7 @@ import java.util.Map;
  */
 public class HazelcastKubernetesDiscoveryStrategyFactory
         implements DiscoveryStrategyFactory {
+    private static final ILogger LOGGER = Logger.getLogger(HazelcastKubernetesDiscoveryStrategyFactory.class);
 
     private static final Collection<PropertyDefinition> PROPERTY_DEFINITIONS;
 
@@ -80,9 +84,25 @@ public class HazelcastKubernetesDiscoveryStrategyFactory
      * @return true if running in the Kubernetes environment
      */
     @Override
-    @SuppressFBWarnings("DMI_HARDCODED_ABSOLUTE_FILENAME")
     public boolean isAutoDetectionApplicable() {
+        return tokenFileExists() && defaultKubernetesMasterReachable();
+    }
+
+    @SuppressFBWarnings("DMI_HARDCODED_ABSOLUTE_FILENAME")
+    private boolean tokenFileExists() {
         return new File("/var/run/secrets/kubernetes.io/serviceaccount/token").exists();
+    }
+
+    private boolean defaultKubernetesMasterReachable() {
+        try {
+            InetAddress.getByName("kubernetes.default.svc");
+            return true;
+        } catch (UnknownHostException e) {
+            LOGGER.warning("Hazelcast running on Kubernetes, but \"kubernetes.default.svc\" is not reachable. "
+                    + "Check your Kubernetes DNS configuration.");
+            LOGGER.finest(e);
+            return false;
+        }
     }
 
     @Override

--- a/src/main/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategyFactory.java
+++ b/src/main/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategyFactory.java
@@ -37,8 +37,6 @@ import java.util.Map;
  */
 public class HazelcastKubernetesDiscoveryStrategyFactory
         implements DiscoveryStrategyFactory {
-    private static final ILogger LOGGER = Logger.getLogger(HazelcastKubernetesDiscoveryStrategyFactory.class);
-
     private static final Collection<PropertyDefinition> PROPERTY_DEFINITIONS;
 
     static {
@@ -98,9 +96,10 @@ public class HazelcastKubernetesDiscoveryStrategyFactory
             InetAddress.getByName("kubernetes.default.svc");
             return true;
         } catch (UnknownHostException e) {
-            LOGGER.warning("Hazelcast running on Kubernetes, but \"kubernetes.default.svc\" is not reachable. "
+            ILogger logger = Logger.getLogger(HazelcastKubernetesDiscoveryStrategyFactory.class);
+            logger.warning("Hazelcast running on Kubernetes, but \"kubernetes.default.svc\" is not reachable. "
                     + "Check your Kubernetes DNS configuration.");
-            LOGGER.finest(e);
+            logger.finest(e);
             return false;
         }
     }

--- a/src/main/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategyFactory.java
+++ b/src/main/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategyFactory.java
@@ -89,7 +89,7 @@ public class HazelcastKubernetesDiscoveryStrategyFactory
     }
 
     @SuppressFBWarnings("DMI_HARDCODED_ABSOLUTE_FILENAME")
-    private boolean tokenFileExists() {
+    boolean tokenFileExists() {
         return new File("/var/run/secrets/kubernetes.io/serviceaccount/token").exists();
     }
 

--- a/src/test/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategyFactoryTest.java
+++ b/src/test/java/com/hazelcast/kubernetes/HazelcastKubernetesDiscoveryStrategyFactoryTest.java
@@ -98,7 +98,7 @@ public class HazelcastKubernetesDiscoveryStrategyFactoryTest {
         HazelcastKubernetesDiscoveryStrategyFactory factory = new HazelcastKubernetesDiscoveryStrategyFactory();
 
         // when & then
-        assertTrue(factory.isAutoDetectionApplicable());
+        assertTrue(factory.tokenFileExists());
         assertEquals(DiscoveryStrategyLevel.PLATFORM, factory.discoveryStrategyLevel());
     }
 }

